ntex-bytes/src/buf/pool.rs

```diff
@@ -0,0 +1,269 @@
+
+    /// The type of backing buffer that is returned to this pool when the last owned or shared buffer
+    /// holding on to said backing buffer is dropped.
+    type Backing;
+
+    /// Returns the given backing buffer to the pool. This is intended to be invoked from
+    /// the `Drop` impls of this pool's owned and shared buffer types.
+    fn put_back(&self, backing: Self::Backing);
+}
+
+/// Convenience trait for a type that is a buffer pool, ie a type that is both a [`BufferSource`]
+/// and a [`BufferSink`].
+pub trait BufferPool: BufferSource + BufferSink<Owned = <Self as BufferSource>::Owned> {}
+
+impl<T> BufferPool for T where T: BufferSource + BufferSink<Owned = <Self as BufferSource>::Owned> {}
+
+impl<T> BufferSink for Rc<T>
+where
+    T: ?Sized + BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+impl<T> BufferSink for Arc<T>
+where
+    T: BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+/// A [`BufferSource`] that always yields new `BytesMut` without blocking.
+#[derive(Clone, Copy, Debug)]
+pub struct BytesPool;
+
+impl BufferSource for BytesPool {
+    type Owned = BytesMut;
+    type Accumulator = BytesAccumulator;
+    type TakeFuture = future::Ready<Self::Owned>;
+
+    fn take(&self, len: usize) -> Self::TakeFuture {
+        future::ready(BytesMut::with_capacity(len))
+    }
+}
+
+impl BufferSink for BytesPool {
+    type Backing = ();
+    type Owned = <Self as BufferSource>::Owned;
+
+    fn put_back(&self, _backing: Self::Backing) {
+    }
+}
+
+/// A buffer that owns a particular range of its backing buffer.
+///
+/// An `Owned` tracks what part of itself has been filled with data.
+/// The filled region can be accessed with [`Owned::filled`].
+/// The unfilled region can be accessed with [`Owned::unfilled_mut`], and bytes can be moved from the start of this region
+/// into the end of the filled region with [`Owned::fill`].
+///
+/// An `Owned` can be subdivided into smaller `Owned`s with `split_to` that each own
+/// smaller splits of the backing buffer.
+///
+/// An `Owned` is not `Clone`. It can be converted to a [`Shared`] which is, via [`Owned::freeze`]
+pub trait Owned {
+    /// The type of shared buffers with the same backing buffer.
+    type Shared: Shared;
+
+    /// Try to append a `u8` to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u8`.
+    fn try_put_u8(&mut self, n: u8) -> Option<()> {
+        self.try_put_slice(&[n])
+    }
+
+    /// Try to append a `u16` in big-endian order to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u16`.
+    fn try_put_u16_be(&mut self, n: u16) -> Option<()> {
```

Comment:
should we always use big endian or network endian and drop `_be` postfixes

Reply:
I'd rather be explicit. It means nobody has to look up the docs to see which one it is, even if there is strong reason to assume it's BE by default.

Reply:
Buf/BufMut does not use _be